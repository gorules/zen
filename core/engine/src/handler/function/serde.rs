use crate::handler::function::error::ResultExt;
use itertools::Itertools;
use rquickjs::{Ctx, FromJs, IntoAtom, IntoJs, Type, Value as QValue};
use serde_json::{json, Map, Number, Value};

#[derive(Debug)]
pub(crate) struct JsValue(pub(crate) Value);

impl<'js> FromJs<'js> for JsValue {
    fn from_js(ctx: &Ctx<'js>, v: QValue<'js>) -> rquickjs::Result<Self> {
        let computed_value = match v.type_of() {
            Type::Uninitialized | Type::Undefined | Type::Null => Value::Null,
            Type::Bool => Value::Bool(v.as_bool().or_throw_msg(ctx, "failed to convert to bool")?),
            Type::Int => Value::Number(Number::from(
                v.as_int().or_throw_msg(ctx, "failed to convert to int")?,
            )),
            Type::BigInt => Value::Number(Number::from(
                v.as_big_int()
                    .map(|b| b.clone().to_i64().ok())
                    .flatten()
                    .or_throw_msg(ctx, "failed to convert to number")?,
            )),
            Type::Float => Value::Number(
                v.as_float()
                    .map(|n| Number::from_f64(n))
                    .flatten()
                    .or_throw_msg(ctx, "failed to convert to number")?,
            ),
            Type::String => Value::String(
                v.as_string()
                    .map(|s| s.to_string().ok())
                    .flatten()
                    .or_throw_msg(ctx, "failed to convert to string")?,
            ),
            Type::Array => {
                let arr = v
                    .as_array()
                    .or_throw_msg(ctx, "failed to convert to array")?;

                let js_arr: Vec<Value> = arr
                    .iter::<QValue>()
                    .map_ok(|n| JsValue::from_js(ctx, n.clone()).map(|js_val| js_val.0))
                    .flatten()
                    .try_collect()
                    .or_throw(ctx)?;

                Value::Array(js_arr)
            }
            Type::Object => {
                let object = v
                    .as_object()
                    .or_throw_msg(ctx, "failed to convert to object")?;

                let js_object: Map<String, Value> = object
                    .props::<String, QValue>()
                    .map_ok(|(key, value)| {
                        JsValue::from_js(ctx, value.clone()).map(|js_val| (key, js_val.0))
                    })
                    .flatten()
                    .try_collect()
                    .or_throw(ctx)?;

                Value::Object(js_object)
            }
            Type::Exception => {
                let exception = v
                    .as_exception()
                    .or_throw_msg(ctx, "failed to convert to exception")?;

                let message = exception.message().unwrap_or_default();
                let description = exception.to_string();

                json!({ "message": message, "description": description })
            }
            Type::Function => json!("[Function]"),
            Type::Module => json!("[Module]"),
            Type::Constructor => json!("[Constructor]"),
            Type::Symbol => json!("[Symbol]"),
            Type::Unknown => json!("[Unknown]"),
            Type::Promise => {
                let promise = v.as_promise().or_throw(ctx)?;
                let val: JsValue = promise.finish()?;
                val.0
            }
        };

        Ok(JsValue(computed_value))
    }
}

impl<'js> IntoJs<'js> for JsValue {
    fn into_js(self, ctx: &Ctx<'js>) -> rquickjs::Result<QValue<'js>> {
        let res = match self.0 {
            Value::Null => QValue::new_null(ctx.clone()),
            Value::Bool(b) => QValue::new_bool(ctx.clone(), b),
            Value::Number(n) => QValue::new_number(
                ctx.clone(),
                n.as_f64()
                    .or_throw_msg(ctx, "failed to convert float to number")?,
            ),
            Value::String(str) => str.into_js(ctx)?,
            Value::Array(arr) => {
                let qarr = rquickjs::Array::new(ctx.clone())?;
                for (idx, item) in arr.into_iter().enumerate() {
                    qarr.set(idx, JsValue(item))?;
                }

                qarr.into_value()
            }
            Value::Object(map) => {
                let qmap = rquickjs::Object::new(ctx.clone())?;
                for (key, value) in map.into_iter() {
                    qmap.set(key.into_atom(ctx)?, JsValue(value))?;
                }

                qmap.into_value()
            }
        };

        Ok(res)
    }
}
