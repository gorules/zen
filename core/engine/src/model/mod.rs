use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[cfg(feature = "bincode")]
use bincode::{Decode, Encode};

/// JDM Decision model
#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionContent {
    pub nodes: Vec<DecisionNode>,
    pub edges: Vec<DecisionEdge>,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionEdge {
    pub source_id: String,
    pub target_id: String,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionNode {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    #[serde(flatten)]
    pub kind: DecisionNodeKind,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(tag = "type")]
#[serde(rename_all = "camelCase")]
pub enum DecisionNodeKind {
    InputNode,
    OutputNode,
    FunctionNode { content: String },
    DecisionNode { content: DecisionNodeContent },
    DecisionTableNode { content: DecisionTableContent },
    ExpressionNode { content: ExpressionNodeContent },
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionNodeContent {
    pub key: String,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionTableContent {
    pub rules: Vec<HashMap<String, String>>,
    pub inputs: Vec<DecisionTableInputField>,
    pub outputs: Vec<DecisionTableOutputField>,
    pub hit_policy: DecisionTableHitPolicy,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub enum DecisionTableHitPolicy {
    First,
    Collect,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionTableInputField {
    pub id: String,
    pub name: String,
    pub field: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct DecisionTableOutputField {
    pub id: String,
    pub name: String,
    pub field: String,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct ExpressionNodeContent {
    pub expressions: Vec<Expression>,
}

#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
#[cfg_attr(feature = "bincode", derive(Encode, Decode))]
#[serde(rename_all = "camelCase")]
pub struct Expression {
    pub id: String,
    pub key: String,
    pub value: String,
}
