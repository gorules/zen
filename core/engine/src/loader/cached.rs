use std::collections::HashMap;
use std::future::Future;
use std::sync::Arc;

use tokio::sync::Mutex;

use crate::loader::{DecisionLoader, LoaderResponse};
use crate::model::DecisionContent;

pub struct CachedLoader<Loader: DecisionLoader + 'static> {
    loader: Arc<Loader>,
    cache: Mutex<HashMap<String, Arc<DecisionContent>>>,
}

impl<Loader: DecisionLoader + 'static> From<Arc<Loader>> for CachedLoader<Loader> {
    fn from(value: Arc<Loader>) -> Self {
        Self {
            loader: value,
            cache: Mutex::new(HashMap::new()),
        }
    }
}

impl<Loader: DecisionLoader + 'static> DecisionLoader for CachedLoader<Loader> {
    fn load<'a>(&'a self, key: &'a str) -> impl Future<Output = LoaderResponse> + 'a {
        async move {
            let mut cache = self.cache.lock().await;
            if let Some(content) = cache.get(key) {
                return Ok(content.clone());
            }

            let decision_content = self.loader.load(key).await?;
            cache.insert(key.to_string(), decision_content.clone());
            Ok(decision_content)
        }
    }
}
