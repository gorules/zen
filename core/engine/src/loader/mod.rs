mod closure;
mod filesystem;
mod memory;
mod noop;

pub use closure::ClosureLoader;
pub use filesystem::{FilesystemLoader, FilesystemLoaderOptions};
pub use memory::MemoryLoader;
pub use noop::NoopLoader;

use async_trait::async_trait;

use crate::model::DecisionContent;
use std::fmt::Debug;
use std::sync::Arc;
use thiserror::Error;

pub type LoaderResult<T> = Result<T, Box<LoaderError>>;
pub type LoaderResponse = LoaderResult<Arc<DecisionContent>>;

/// Trait used for implementing a loader for decisions
#[async_trait]
pub trait DecisionLoader {
    async fn load(&self, key: &str) -> LoaderResponse;
}

#[derive(Error, Debug)]
pub enum LoaderError {
    #[error("Loader did not find item with key {0}")]
    NotFound(String),
    #[error("Loader failed internally on key {key}: {source}.")]
    Internal {
        key: String,
        #[source]
        source: anyhow::Error,
    },
}
