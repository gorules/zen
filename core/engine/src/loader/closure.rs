use crate::loader::{DecisionLoader, LoaderResponse};
use async_trait::async_trait;
use std::future::Future;

/// Loads decisions using an async closure
#[derive(Debug)]
pub struct ClosureLoader<F>
where
    F: Sync + Send,
{
    closure: F,
}

impl<F, O> ClosureLoader<F>
where
    F: Fn(&str) -> O + Sync + Send,
    O: Future<Output = LoaderResponse> + Send,
{
    pub fn new(closure: F) -> Self {
        Self { closure }
    }
}

#[async_trait]
impl<F, O> DecisionLoader for ClosureLoader<F>
where
    F: Fn(&str) -> O + Sync + Send,
    O: Future<Output = LoaderResponse> + Send,
{
    async fn load(&self, key: &str) -> LoaderResponse {
        let closure = &self.closure;
        closure(key).await
    }
}
