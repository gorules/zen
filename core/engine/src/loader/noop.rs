use crate::loader::{DecisionLoader, LoaderError, LoaderResponse};
use anyhow::anyhow;
use async_trait::async_trait;

/// Default loader which always fails
#[derive(Default, Debug)]
pub struct NoopLoader;

#[async_trait]
impl DecisionLoader for NoopLoader {
    async fn load(&self, key: &str) -> LoaderResponse {
        Err(LoaderError::Internal {
            key: key.to_string(),
            source: anyhow!("Loader is no-op"),
        }
        .into())
    }
}
