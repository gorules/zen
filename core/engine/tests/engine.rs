use std::io::Read;
use std::ops::Deref;
use std::path::Path;
use std::sync::Arc;

use serde::Deserialize;
use serde_json::json;

use zen_engine::loader::{LoaderError, MemoryLoader};
use zen_engine::model::DecisionNodeKind;
use zen_engine::{DecisionEngine, EvaluationError, EvaluationOptions};

use crate::support::{create_fs_loader, load_raw_test_data, load_test_data};

mod support;

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_memory_loader() {
    let memory_loader = Arc::new(MemoryLoader::default());
    memory_loader.add("table", load_test_data("table.json"));
    memory_loader.add("function", load_test_data("function.json"));

    let engine = DecisionEngine::new_arc(memory_loader.clone());
    let table = engine.evaluate("table", &json!({ "input": 12 })).await;
    let function = engine.evaluate("function", &json!({ "input": 12 })).await;

    memory_loader.remove("function");
    let not_found = engine.evaluate("function", &json!({})).await;

    assert_eq!(table.unwrap().result, json!({"output": 10}));
    assert_eq!(function.unwrap().result, json!({"output": 24}));
    assert_eq!(not_found.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_filesystem_loader() {
    let engine = DecisionEngine::new(create_fs_loader());
    let table = engine.evaluate("table.json", &json!({ "input": 12 })).await;
    let function = engine
        .evaluate("function.json", &json!({ "input": 12 }))
        .await;
    let not_found = engine.evaluate("invalid_file", &json!({})).await;

    assert_eq!(table.unwrap().result, json!({"output": 10}));
    assert_eq!(function.unwrap().result, json!({"output": 24}));
    assert_eq!(not_found.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_closure_loader() {
    let engine = DecisionEngine::async_loader(|key| {
        // TODO: Improve once async closures become stable in Rust
        let mv_key = key.to_string();

        async move {
            match mv_key.as_str() {
                "function" => Ok(Arc::new(load_test_data("function.json"))),
                "table" => Ok(Arc::new(load_test_data("table.json"))),
                _ => Err(LoaderError::NotFound(mv_key).into()),
            }
        }
    });

    let table = engine.evaluate("table", &json!({ "input": 12 })).await;
    let function = engine.evaluate("function", &json!({ "input": 12 })).await;
    let not_found = engine.evaluate("invalid_file", &json!({})).await;

    assert_eq!(table.unwrap().result, json!({"output": 10}));
    assert_eq!(function.unwrap().result, json!({"output": 24}));
    assert_eq!(not_found.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
async fn engine_noop_loader() {
    // Default engine is noop
    let engine = DecisionEngine::default();
    let result = engine.evaluate("any.json", &json!({})).await;

    assert_eq!(result.unwrap_err().to_string(), "Loader error");
}

#[tokio::test]
async fn engine_get_decision() {
    let engine = DecisionEngine::new(create_fs_loader());

    assert!(engine.get_decision("table.json").await.is_ok());
    assert!(engine.get_decision("any.json").await.is_err());
}

#[tokio::test]
async fn engine_create_decision() {
    let engine = DecisionEngine::default();
    engine.create_decision(load_test_data("table.json").into());
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_errors() {
    let engine = DecisionEngine::new(create_fs_loader());

    let infinite_fn = engine.evaluate("infinite-function.json", &json!({})).await;
    match infinite_fn.unwrap_err().deref() {
        EvaluationError::NodeError(e) => {
            assert_eq!(e.node_id, "e0fd96d0-44dc-4f0e-b825-06e56b442d78");
            assert_eq!(e.source.to_string(), "Timeout exceeded");
        }
        _ => assert!(false, "Wrong error type"),
    }

    let recursive = engine.evaluate("recursive-table1.json", &json!({})).await;
    match recursive.unwrap_err().deref() {
        EvaluationError::NodeError(e) => {
            assert_eq!(e.source.to_string(), "Depth limit exceeded")
        }
        _ => assert!(false, "Depth limit not exceeded"),
    }
}

#[tokio::test]
async fn engine_with_trace() {
    let engine = DecisionEngine::new(create_fs_loader());

    let table_r = engine.evaluate("table.json", &json!({ "input": 12 })).await;
    let table_opt_r = engine
        .evaluate_with_opts(
            "table.json",
            &json!({ "input": 12 }),
            EvaluationOptions {
                trace: Some(true),
                max_depth: None,
            },
        )
        .await;

    let table = table_r.unwrap();
    let table_opt = table_opt_r.unwrap();

    assert!(table.trace.is_none());
    assert!(table_opt.trace.is_some());

    let trace = table_opt.trace.unwrap();
    assert_eq!(trace.len(), 3); // trace for each node
}

#[tokio::test]
#[cfg_attr(miri, ignore)]
async fn engine_function_imports() {
    let mut function_content = load_test_data("function.json");

    let imports_js_path = Path::new("js").join("imports.js");
    let mut replace_buffer = load_raw_test_data(imports_js_path.to_str().unwrap());
    let mut replace_data = String::new();
    replace_buffer.read_to_string(&mut replace_data).unwrap();

    function_content.nodes.iter_mut().for_each(|node| {
        if let DecisionNodeKind::FunctionNode { content } = &mut node.kind {
            let _ = std::mem::replace(content, replace_data.clone());
        }
    });

    let decision = DecisionEngine::default().create_decision(function_content.into());
    let response = decision.evaluate(&json!({})).await.unwrap();

    #[derive(Deserialize, Debug)]
    #[serde(rename_all = "camelCase")]
    struct GraphResult {
        bigjs_tests: Vec<bool>,
        bigjs_valid: bool,
        dayjs_valid: bool,
        moment_valid: bool,
    }

    let result = serde_json::from_value::<GraphResult>(response.result).unwrap();

    assert!(result.bigjs_tests.iter().all(|v| *v));
    assert!(result.bigjs_valid);
    assert!(result.dayjs_valid);
    assert!(result.moment_valid);
}
