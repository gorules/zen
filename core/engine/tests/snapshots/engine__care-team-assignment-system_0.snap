---
source: core/engine/tests/engine.rs
expression: serialized_result
---
performance: "[perf]"
result:
  ageCategory: senior
  careTeam:
    complexityLevel: high
    coordinator:
      priority: high
      role: care_coordinator
    providers:
      - priority: high
        role: primary_physician
      - priority: medium
        role: geriatric_specialist
      - priority: medium
        role: mental_health_provider
      - priority: medium
        role: nutritionist
      - priority: high
        role: nurse_practitioner
      - priority: medium
        role: physical_therapist
    teamSize: "3"
  mobilityIssues: true
  requiresMentalHealthSupport: true
  requiresNutritionSupport: true
  requiresSpecialist: true
trace:
  dt1:
    id: dt1
    input:
      availableProviders:
        - id: DR001
          languages:
            - english
            - spanish
          name: Dr. Sarah Johnson
          role: primary_physician
          specialties:
            - internal_medicine
            - geriatrics
        - id: DR002
          languages:
            - english
            - mandarin
          name: Dr. Michael Chen
          role: cardiologist
          specialties:
            - cardiology
        - id: DR003
          languages:
            - english
          name: Dr. Lisa Peterson
          role: diabetes_specialist
          specialties:
            - endocrinology
        - id: NP001
          languages:
            - english
          name: Nancy Williams
          role: nurse_practitioner
          specialties:
            - primary_care
        - id: RN001
          languages:
            - english
            - spanish
          name: Robert Garcia
          role: nurse
          specialties:
            - geriatric_care
        - id: MH001
          languages:
            - english
          name: Dr. Emily Cohen
          role: mental_health_provider
          specialties:
            - psychiatry
            - geriatric_psychiatry
        - id: PT001
          languages:
            - english
          name: James Wilson
          role: physical_therapist
          specialties:
            - geriatric_physical_therapy
        - id: CC001
          languages:
            - english
            - spanish
          name: Maria Rodriguez
          role: care_coordinator
          specialties:
            - complex_care_coordination
      facility:
        department: Internal Medicine
        id: FAC789
        name: Memorial Medical Center
      patient:
        age:
          "$serde_json::private::Number": "67"
        careComplexity: high
        carePreferences:
          preferredGender: any
          preferredLanguage: english
          requiresTranslator: false
        conditions:
          - diabetes
          - heart_disease
          - hypertension
          - depression
        gender: male
        id: PT12345
        insuranceType: medicare
        mobilityScore:
          "$serde_json::private::Number": "35"
        name: John Smith
        primaryLanguage: english
    name: determine_complexity
    order:
      "$serde_json::private::Number": "1"
    output:
      availableProviders:
        - id: DR001
          languages:
            - english
            - spanish
          name: Dr. Sarah Johnson
          role: primary_physician
          specialties:
            - internal_medicine
            - geriatrics
        - id: DR002
          languages:
            - english
            - mandarin
          name: Dr. Michael Chen
          role: cardiologist
          specialties:
            - cardiology
        - id: DR003
          languages:
            - english
          name: Dr. Lisa Peterson
          role: diabetes_specialist
          specialties:
            - endocrinology
        - id: NP001
          languages:
            - english
          name: Nancy Williams
          role: nurse_practitioner
          specialties:
            - primary_care
        - id: RN001
          languages:
            - english
            - spanish
          name: Robert Garcia
          role: nurse
          specialties:
            - geriatric_care
        - id: MH001
          languages:
            - english
          name: Dr. Emily Cohen
          role: mental_health_provider
          specialties:
            - psychiatry
            - geriatric_psychiatry
        - id: PT001
          languages:
            - english
          name: James Wilson
          role: physical_therapist
          specialties:
            - geriatric_physical_therapy
        - id: CC001
          languages:
            - english
            - spanish
          name: Maria Rodriguez
          role: care_coordinator
          specialties:
            - complex_care_coordination
      careTeam:
        complexityLevel: high
        teamSize: "3"
      facility:
        department: Internal Medicine
        id: FAC789
        name: Memorial Medical Center
      patient:
        age:
          "$serde_json::private::Number": "67"
        careComplexity: high
        carePreferences:
          preferredGender: any
          preferredLanguage: english
          requiresTranslator: false
        conditions:
          - diabetes
          - heart_disease
          - hypertension
          - depression
        gender: male
        id: PT12345
        insuranceType: medicare
        mobilityScore:
          "$serde_json::private::Number": "35"
        name: John Smith
        primaryLanguage: english
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "0"
      reference_map:
        patient.careComplexity: high
      rule:
        _id: r1-1
        "patient.careComplexity[i1-1]": "\"high\""
  dt2:
    id: dt2
    input:
      ageCategory: senior
      careTeam:
        complexityLevel: high
        teamSize: "3"
      mobilityIssues: true
      requiresMentalHealthSupport: true
      requiresNutritionSupport: true
      requiresSpecialist: true
    name: assign_providers
    order:
      "$serde_json::private::Number": "4"
    output:
      ageCategory: senior
      careTeam:
        complexityLevel: high
        providers:
          - priority: high
            role: primary_physician
          - priority: medium
            role: geriatric_specialist
          - priority: medium
            role: mental_health_provider
          - priority: medium
            role: nutritionist
          - priority: high
            role: nurse_practitioner
          - priority: medium
            role: physical_therapist
        teamSize: "3"
      mobilityIssues: true
      requiresMentalHealthSupport: true
      requiresNutritionSupport: true
      requiresSpecialist: true
    performance: "[perf]"
    traceData:
      - index:
          "$serde_json::private::Number": "0"
        reference_map:
          careTeam.complexityLevel: high
        rule:
          _id: r2-1
          "careTeam.complexityLevel[i2-1]": "\"high\""
      - index:
          "$serde_json::private::Number": "2"
        reference_map:
          careTeam.complexityLevel: high
        rule:
          _id: r2-3
          "careTeam.complexityLevel[i2-1]": ""
      - index:
          "$serde_json::private::Number": "7"
        reference_map:
          careTeam.complexityLevel: high
        rule:
          _id: r2-8
          "careTeam.complexityLevel[i2-1]": ""
      - index:
          "$serde_json::private::Number": "8"
        reference_map:
          careTeam.complexityLevel: high
        rule:
          _id: r2-9
          "careTeam.complexityLevel[i2-1]": ""
      - index:
          "$serde_json::private::Number": "9"
        reference_map:
          careTeam.complexityLevel: high
        rule:
          _id: r2-10
          "careTeam.complexityLevel[i2-1]": "\"high\""
      - index:
          "$serde_json::private::Number": "11"
        reference_map:
          careTeam.complexityLevel: high
        rule:
          _id: r2-12
          "careTeam.complexityLevel[i2-1]": ""
  dt3:
    id: dt3
    input:
      ageCategory: senior
      careTeam:
        complexityLevel: high
        providers:
          - priority: high
            role: primary_physician
          - priority: medium
            role: geriatric_specialist
          - priority: medium
            role: mental_health_provider
          - priority: medium
            role: nutritionist
          - priority: high
            role: nurse_practitioner
          - priority: medium
            role: physical_therapist
        teamSize: "3"
      mobilityIssues: true
      requiresMentalHealthSupport: true
      requiresNutritionSupport: true
      requiresSpecialist: true
    name: assign_coordinator
    order:
      "$serde_json::private::Number": "5"
    output:
      ageCategory: senior
      careTeam:
        complexityLevel: high
        coordinator:
          priority: high
          role: care_coordinator
        providers:
          - priority: high
            role: primary_physician
          - priority: medium
            role: geriatric_specialist
          - priority: medium
            role: mental_health_provider
          - priority: medium
            role: nutritionist
          - priority: high
            role: nurse_practitioner
          - priority: medium
            role: physical_therapist
        teamSize: "3"
      mobilityIssues: true
      requiresMentalHealthSupport: true
      requiresNutritionSupport: true
      requiresSpecialist: true
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "0"
      reference_map:
        $: high
        careTeam.complexityLevel: high
      rule:
        "$[i3-2]": len(careTeam.providers) > 5
        _id: r3-1
        "careTeam.complexityLevel[i3-1]": "\"high\""
  ex1:
    id: ex1
    input:
      ageCategory: senior
      availableProviders:
        - id: DR001
          languages:
            - english
            - spanish
          name: Dr. Sarah Johnson
          role: primary_physician
          specialties:
            - internal_medicine
            - geriatrics
        - id: DR002
          languages:
            - english
            - mandarin
          name: Dr. Michael Chen
          role: cardiologist
          specialties:
            - cardiology
        - id: DR003
          languages:
            - english
          name: Dr. Lisa Peterson
          role: diabetes_specialist
          specialties:
            - endocrinology
        - id: NP001
          languages:
            - english
          name: Nancy Williams
          role: nurse_practitioner
          specialties:
            - primary_care
        - id: RN001
          languages:
            - english
            - spanish
          name: Robert Garcia
          role: nurse
          specialties:
            - geriatric_care
        - id: MH001
          languages:
            - english
          name: Dr. Emily Cohen
          role: mental_health_provider
          specialties:
            - psychiatry
            - geriatric_psychiatry
        - id: PT001
          languages:
            - english
          name: James Wilson
          role: physical_therapist
          specialties:
            - geriatric_physical_therapy
        - id: CC001
          languages:
            - english
            - spanish
          name: Maria Rodriguez
          role: care_coordinator
          specialties:
            - complex_care_coordination
      careTeam:
        complexityLevel: high
        teamSize: "3"
      facility:
        department: Internal Medicine
        id: FAC789
        name: Memorial Medical Center
      patient:
        age:
          "$serde_json::private::Number": "67"
        careComplexity: high
        carePreferences:
          preferredGender: any
          preferredLanguage: english
          requiresTranslator: false
        conditions:
          - diabetes
          - heart_disease
          - hypertension
          - depression
        gender: male
        id: PT12345
        insuranceType: medicare
        mobilityScore:
          "$serde_json::private::Number": "35"
        name: John Smith
        primaryLanguage: english
    name: evaluate_patient_needs
    order:
      "$serde_json::private::Number": "3"
    output:
      ageCategory: senior
      careTeam:
        complexityLevel: high
        teamSize: "3"
      mobilityIssues: true
      requiresMentalHealthSupport: true
      requiresNutritionSupport: true
      requiresSpecialist: true
    performance: "[perf]"
    traceData:
      availableProviders:
        result: ~
      facility:
        result: ~
      mobilityIssues:
        result: true
      patient:
        result: ~
      requiresMentalHealthSupport:
        result: true
      requiresNutritionSupport:
        result: true
      requiresSpecialist:
        result: true
  f1af25e2-126e-4882-8d35-6ae33dbc37bf:
    id: f1af25e2-126e-4882-8d35-6ae33dbc37bf
    input:
      availableProviders:
        - id: DR001
          languages:
            - english
            - spanish
          name: Dr. Sarah Johnson
          role: primary_physician
          specialties:
            - internal_medicine
            - geriatrics
        - id: DR002
          languages:
            - english
            - mandarin
          name: Dr. Michael Chen
          role: cardiologist
          specialties:
            - cardiology
        - id: DR003
          languages:
            - english
          name: Dr. Lisa Peterson
          role: diabetes_specialist
          specialties:
            - endocrinology
        - id: NP001
          languages:
            - english
          name: Nancy Williams
          role: nurse_practitioner
          specialties:
            - primary_care
        - id: RN001
          languages:
            - english
            - spanish
          name: Robert Garcia
          role: nurse
          specialties:
            - geriatric_care
        - id: MH001
          languages:
            - english
          name: Dr. Emily Cohen
          role: mental_health_provider
          specialties:
            - psychiatry
            - geriatric_psychiatry
        - id: PT001
          languages:
            - english
          name: James Wilson
          role: physical_therapist
          specialties:
            - geriatric_physical_therapy
        - id: CC001
          languages:
            - english
            - spanish
          name: Maria Rodriguez
          role: care_coordinator
          specialties:
            - complex_care_coordination
      careTeam:
        complexityLevel: high
        teamSize: "3"
      facility:
        department: Internal Medicine
        id: FAC789
        name: Memorial Medical Center
      patient:
        age:
          "$serde_json::private::Number": "67"
        careComplexity: high
        carePreferences:
          preferredGender: any
          preferredLanguage: english
          requiresTranslator: false
        conditions:
          - diabetes
          - heart_disease
          - hypertension
          - depression
        gender: male
        id: PT12345
        insuranceType: medicare
        mobilityScore:
          "$serde_json::private::Number": "35"
        name: John Smith
        primaryLanguage: english
    name: age_category
    order:
      "$serde_json::private::Number": "2"
    output:
      ageCategory: senior
      availableProviders:
        - id: DR001
          languages:
            - english
            - spanish
          name: Dr. Sarah Johnson
          role: primary_physician
          specialties:
            - internal_medicine
            - geriatrics
        - id: DR002
          languages:
            - english
            - mandarin
          name: Dr. Michael Chen
          role: cardiologist
          specialties:
            - cardiology
        - id: DR003
          languages:
            - english
          name: Dr. Lisa Peterson
          role: diabetes_specialist
          specialties:
            - endocrinology
        - id: NP001
          languages:
            - english
          name: Nancy Williams
          role: nurse_practitioner
          specialties:
            - primary_care
        - id: RN001
          languages:
            - english
            - spanish
          name: Robert Garcia
          role: nurse
          specialties:
            - geriatric_care
        - id: MH001
          languages:
            - english
          name: Dr. Emily Cohen
          role: mental_health_provider
          specialties:
            - psychiatry
            - geriatric_psychiatry
        - id: PT001
          languages:
            - english
          name: James Wilson
          role: physical_therapist
          specialties:
            - geriatric_physical_therapy
        - id: CC001
          languages:
            - english
            - spanish
          name: Maria Rodriguez
          role: care_coordinator
          specialties:
            - complex_care_coordination
      careTeam:
        complexityLevel: high
        teamSize: "3"
      facility:
        department: Internal Medicine
        id: FAC789
        name: Memorial Medical Center
      patient:
        age:
          "$serde_json::private::Number": "67"
        careComplexity: high
        carePreferences:
          preferredGender: any
          preferredLanguage: english
          requiresTranslator: false
        conditions:
          - diabetes
          - heart_disease
          - hypertension
          - depression
        gender: male
        id: PT12345
        insuranceType: medicare
        mobilityScore:
          "$serde_json::private::Number": "35"
        name: John Smith
        primaryLanguage: english
    performance: "[perf]"
    traceData:
      index:
        "$serde_json::private::Number": "2"
      reference_map:
        patient.age:
          "$serde_json::private::Number": "67"
      rule:
        _id: r1-3
        "patient.age[i1-1]": ""
  ip1:
    id: ip1
    input: ~
    name: request
    order:
      "$serde_json::private::Number": "0"
    output:
      availableProviders:
        - id: DR001
          languages:
            - english
            - spanish
          name: Dr. Sarah Johnson
          role: primary_physician
          specialties:
            - internal_medicine
            - geriatrics
        - id: DR002
          languages:
            - english
            - mandarin
          name: Dr. Michael Chen
          role: cardiologist
          specialties:
            - cardiology
        - id: DR003
          languages:
            - english
          name: Dr. Lisa Peterson
          role: diabetes_specialist
          specialties:
            - endocrinology
        - id: NP001
          languages:
            - english
          name: Nancy Williams
          role: nurse_practitioner
          specialties:
            - primary_care
        - id: RN001
          languages:
            - english
            - spanish
          name: Robert Garcia
          role: nurse
          specialties:
            - geriatric_care
        - id: MH001
          languages:
            - english
          name: Dr. Emily Cohen
          role: mental_health_provider
          specialties:
            - psychiatry
            - geriatric_psychiatry
        - id: PT001
          languages:
            - english
          name: James Wilson
          role: physical_therapist
          specialties:
            - geriatric_physical_therapy
        - id: CC001
          languages:
            - english
            - spanish
          name: Maria Rodriguez
          role: care_coordinator
          specialties:
            - complex_care_coordination
      facility:
        department: Internal Medicine
        id: FAC789
        name: Memorial Medical Center
      patient:
        age:
          "$serde_json::private::Number": "67"
        careComplexity: high
        carePreferences:
          preferredGender: any
          preferredLanguage: english
          requiresTranslator: false
        conditions:
          - diabetes
          - heart_disease
          - hypertension
          - depression
        gender: male
        id: PT12345
        insuranceType: medicare
        mobilityScore:
          "$serde_json::private::Number": "35"
        name: John Smith
        primaryLanguage: english
    performance: "[perf]"
    traceData: ~
